fix: Exception with disabled GameObjects and NetworkTransforms

EmandM opened the pull request:
<!-- Replace this block with what this PR does and why. Describe what you'd like reviewers to know, how you applied the engineering principles, and any interesting tradeoffs made. Delete bullet points below that don't apply, and update the changelog section as appropriate. -->

<!-- Add short version of the JIRA ticket to the PR title (e.g. "feat: new shiny feature [MTT-123]") -->
[MTTB-959](https://jira.unity3d.com/browse/MTTB-959)

<!-- Add RFC link here if applicable. -->

## Changelog

- Fixed: Log an appropriate error if the user has disabled a `GameObject` associated with a `NetworkBehaviour`. Do not throw an excpetion.

## Testing and Documentation

- Includes unit tests.

<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

EmandM committed:
fix: Exception with disabled GameObjects and NetworkTransforms

EmandM committed:
Add CHANGELOG

EmandM committed:
Merge branch 'develop-2.0.0' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/mttb-959-exception-with-disabled-gameobjects

EmandM committed:
Update naming and code comments to be clearer

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/NetworkTransformMessage.cs:
```diff
@@ -71,8 +71,21 @@ public bool Deserialize(FastBufferReader reader, ref NetworkContext context, int
                 networkObject = networkManager.SpawnManager.SpawnedObjects[networkObjectId];
+                if (networkObject.ChildNetworkBehaviours.Count <= networkBehaviourId || networkObject.ChildNetworkBehaviours[networkBehaviourId] == null)
+                {
+                    Debug.LogError($"[{nameof(NetworkTransformMessage)}][Invalid] Targeted {nameof(NetworkTransform)}, {nameof(NetworkBehaviour.NetworkBehaviourId)} ({networkBehaviourId}), does not exist! Make sure you are not spawning {nameof(NetworkObject)}s with disabled {nameof(GameObject)}s that have {nameof(NetworkBehaviour)} components on them.");
```
Nice wordsmithing...

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/NetworkTransformMessage.cs:
```diff
@@ -71,8 +71,21 @@ public bool Deserialize(FastBufferReader reader, ref NetworkContext context, int
+                var transform = networkObject.ChildNetworkBehaviours[networkBehaviourId] as NetworkTransform;
+                if (transform == null)
+                {
+                    Debug.LogError($"[{nameof(NetworkTransformMessage)}][Invalid] Targeted {nameof(NetworkTransform)}, {nameof(NetworkBehaviour.NetworkBehaviourId)} ({networkBehaviourId}), does not exist! Make sure you are not spawning {nameof(NetworkObject)}s with disabled {nameof(GameObject)}s that have {nameof(NetworkBehaviour)} components on them.");
```
➕

NoelStephensUnity committed:
fix

Minor adjustment as it looks like DAHost might not have been properly forwarding messages of NetworkTransformMessages when the NetworkObject was hidden from it.

NoelStephensUnity committed:
fix

Some minor tweaks for extracting the reliability used.

NoelStephensUnity committed:
style

updating the test LogAssert.Expect string used.

NoelStephensUnity committed:
update + style

Just keeping the established pattern of using "nameof" within a string.
Also added the NetworkTransform's NetworkBehaviourId value as opposed to hard coding the value.
(minor tweaks)

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/NetworkTransformMessage.cs:
```diff
@@ -81,8 +94,20 @@ public bool Deserialize(FastBufferReader reader, ref NetworkContext context, int
             {
-                // Deserialize the state
-                reader.ReadNetworkSerializableInPlace(ref State);
+                ownerAuthoritativeServerSide = networkManager.DAHost;
```
While reviewing your PR, I noticed a bug in DAHost mode.
Just including the fix for that here.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/NetworkTransformMessage.cs:
```diff
@@ -132,7 +151,10 @@ public bool Deserialize(FastBufferReader reader, ref NetworkContext context, int
                     }
 
-                    var networkDelivery = State.IsReliableStateUpdate() ? NetworkDelivery.ReliableSequenced : NetworkDelivery.UnreliableSequenced;
+                    // Depending upon whether it is spawned locally or not, get the deserialized state
```
This is the 2nd part of the fix where State was never being set and it wasn't adjusting based on whether the host, server, or DAHost had a NetworkTransform instance.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformErrorTests.cs:
```diff
@@ -0,0 +1,140 @@
+            yield return WaitForConditionOrTimeOut(() => ObjectSpawnedOnAllClients(networkObjectInstance.NetworkObjectId));
+            AssertOnTimeout("Timed out waiting for object to spawn!");
+
+            var errorMessage = $"[Netcode] {nameof(NetworkBehaviour)} index {networkTransformInstance.NetworkBehaviourId} was out of bounds for {m_NonAuthorityPrefab.name}(Clone). " +
```
Minor adjustments.

NoelStephensUnity reviewed (approved):
Had to add a fix for another issue I noticed while reviewing your PR (not due to your PR but that already existed).

Did a minor cosmetic update to the error message generation for LogAssert.Expect.

Other than that, it looks good to me!

_As long as my fix didn't break something else!_ 😝
